std/zig/parse2.zig

```diff
@@ -0,0 +1,2978 @@
+const std = @import("../std.zig");
+const assert = std.debug.assert;
+const Allocator = std.mem.Allocator;
+const ast = std.zig.ast;
+const Node = ast.Node;
+const Tree = ast.Tree;
+const AstError = ast.Error;
+const TokenIndex = ast.TokenIndex;
+const Token = std.zig.Token;
+const TokenIterator = Tree.TokenList.Iterator;
+
+pub const Error = error{UnexpectedToken} || Allocator.Error;
+
+pub fn parse(allocator: *Allocator, source: []const u8) !*Tree {
+    var tree_arena = std.heap.ArenaAllocator.init(allocator);
+    errdefer tree_arena.deinit();
+    const arena = &tree_arena.allocator;
+
+    var token_list = Tree.TokenList.init(arena);
+    var tokenizer = std.zig.Tokenizer.init(source);
+    while (true) {
+        const tree_token = try token_list.addOne();
+        tree_token.* = tokenizer.next();
+        if (tree_token.id == .Eof) break;
+    }
+    var it = token_list.iterator(0);
+
+    while (it.peek().?.id == .LineComment) _ = it.next();
+
+    const tree = try arena.create(Tree);
+    tree.* = Tree{
+        .source = source,
+        .root_node = undefined,
+        .tokens = token_list,
+        .errors = Tree.ErrorList.init(arena),
+        .arena_allocator = tree_arena,
+    };
+
+    tree.root_node = parseRoot(&tree.arena_allocator.allocator, &it, tree) catch |err| {
+        return switch (err) {
+            Error.UnexpectedToken => tree,
+            else => err,
+        };
+    };
+
+    return tree;
+}
+
+// Root <- skip ContainerMembers eof
+fn parseRoot(arena: *Allocator, it: *TokenIterator, tree: *Tree) !*Node.Root {
+    const node = try arena.create(Node.Root);
+    node.* = Node.Root{
+        .base = Node{ .id = .Root },
+        .decls = undefined,
+        // TODO: File an issue: Currently it is impossible to have a container-level
```

Comment:
Believe this is covered by #2288?

Reply:
Ah, I think you're right @tiehuis -- although it's a variation on the problem I encountered. I tried a top-level doc comment with _no_ comment on the first item, rather than your example of both comments becoming collapsed into one. I think they are both caused by the same problem: `zig fmt` collapses newlines below comments. It seems to me that behavior will have to change in order to fix the problem -- what do you think?

(As an aside, I have bigger changes in mind for doc comments in general, but that will definitely have to wait until the parser rewrite is finished.)